lumicks/pylake/piezo_tracking/piezo_tracking.py

```diff
@@ -0,0 +1,178 @@
+        """
+        trap_position, camera_distance = trap_position.downsampled_like(camera_distance)
+        mask = camera_distance.data != 0
+        missed_frames = np.sum(1 - mask)
+        if missed_frames > 0:
+            warnings.warn(
+                RuntimeWarning(
+                    "There were frames with missing video tracking: "
+                    f"{missed_frames} data point(s) were omitted."
+                )
+            )
+        self.position, self.distance = trap_position.data[mask], camera_distance.data[mask]
+        coeffs = np.polyfit(self.position, self.distance, degree)
+        self._model = np.poly1d(coeffs)
+
+    def __call__(self, trap_position):
+        return Slice(
+            trap_position._src._with_data(self._model(trap_position.data)),
+            labels={"title": "Piezo distance", "y": "Distance [um]"},
+        )
+
+    def valid_range(self):
+        return (np.min(self.position), np.max(self.position))
+
+    def __str__(self):
+        powers = np.flip(np.arange(self._model.order + 1))
+        return "".join(
+            f"{' + ' if coeff > 0 else ' - '}"
+            f"{abs(coeff):.4f}"
+            f"{'' if power == 0 else ' x' if power == 1 else f' x^{power}'}"
+            for power, coeff in zip(powers, self._model.coeffs)
+        ).strip()
+
+    def __repr__(self):
+        return f"DistanceCalibration({str(self)})"
+
+    def plot(self):
+        """Plot the calibration fit"""
+        plt.scatter(self.position, self.distance, s=2, label="data")
+        plt.plot(self.position, self._model(self.position), "k", label=f"${str(self)}$")
+        plt.xlabel("Mirror position")
+        plt.ylabel("Camera Distance [um]")
+        plt.tight_layout()
+        plt.legend()
+
+    def plot_residual(self):
+        """Plot the residual of the calibration fit"""
+        plt.scatter(self.position, self._model(self.position) - self.distance, s=2)
+        plt.ylabel("Residual [um]")
+        plt.xlabel("Mirror position")
+        plt.tight_layout()
+        plt.legend()
+
+    @classmethod
+    def from_file(cls, calibration_file, degree=1):
+        """Use a reference measurement to calibrate trap mirror position to bead-bead distance.
+
+        Parameters
+        ----------
+        calibration_file : pylake.File
+        degree : int
+            Polynomial order.
+        """
+        return cls(calibration_file["Trap position"]["1X"], calibration_file.distance1, degree)
+
+
+class PiezoTrackingCalibration:
+    def __init__(
+        self,
+        trap_calibration,
+        baseline_force1,
+        baseline_force2,
+        signs=(1, -1),
+    ):
+        """Set up piezo tracking calibration
+
+        trap_calibration : pylake.DistanceCalibration
+            Calibration from trap position to trap to trap distance.
+        baseline_force1 : pylake.ForceBaseline
+            Baseline for force1
+        baseline_force2 : pylake.ForceBaseline
+            Baseline for force2
```

Comment:
I don't really think it's optional. Even if a trap is stationary, its baseline varies as the other trap moves closer. It's actually often the static trap whose baseline corrected force you end up using.

That said, I can see why you'd want to decouple the two. Correcting the force is technically indeed not part of piezo tracking.

However, I'm a bit worried about the potential for user error if we take it out. In most cases, you want to grab F,d relations (either state transitions or an F,d curve). Ideally, we'd need a way ensure that the points we get from the force are still correlated with the ones we get from the distance. Since both potentially have different validity limits (due to being calibrated from different data), they may have different validity ranges.

How would you feel about taking it out of this class, but then making a different convenience class that does both for the user (and uses this one for the piezo tracking part)?

Reply:
Making them optional seems like a good compromise. I guess we'll still need to do some major revisions to the online tracking.